src/x86_64/kvm_platform.c

```diff
@@ -21,3 +39,56 @@ void halt(char *format, ...)
     vm_exit(VM_EXIT_HALT);
 }
 
+static boolean probe_kvm_pvclock(kernel_heaps kh)
+{
+    kvm_debug("probing for KVM pvclock...");
+    heap backed = heap_backed(kh);
+    u32 v[4];
+    cpuid(KVM_CPUID_FEATURES, 0, v);
+    if ((v[0] & (1 << 3)) == 0) {
+        kvm_debug("no pvclock detected");
+        return false;
+    }
+    kvm_debug("pvclock detected");
+    struct pvclock_vcpu_time_info * vc = allocate(backed, backed->pagesize);
+    zero(vc, sizeof(struct pvclock_vcpu_time_info));
+    kvm_debug("before write msr");
+    write_msr(KVM_MSR_SYSTEM_TIME, physical_from_virtual(vc) | /* enable */ 1);
+    memory_barrier();
+    kvm_debug("after write msr");
+    if (vc->system_time == 0) {
+        /* noise, but we should know if this happens */
+        msg_err("kvm pvclock probe failed: detected kvm pvclock, but system_time == 0\n");
+        return false;
+    }
+    init_pvclock(heap_general(kh), vc);
+    return true;
+}
+
+extern void lapic_runloop_timer(timestamp interval);
```

Comment:
For this and for other apic functions:
`configure_lapic_timer`
`lapic_eoi`
`init_apic`
 .... would it make sense to refactor into apic.h?

Reply:
Fixed.